src/Orleans/Telemetry/TelemetryManager.cs

```diff
@@ -0,0 +1,259 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+using Orleans.Runtime.Configuration;
+
+namespace Orleans.Runtime
+{
+    public class TelemetryManager : ITelemetryClient, IDisposable
+    {
+        internal const string ObsoleteMessageTelemetry = "This method might be removed in the future in favor of a non Orleans-owned abstraction for APMs.";
+
+        private List<ITelemetryConsumer> consumers = new List<ITelemetryConsumer>();
+        private List<IMetricTelemetryConsumer> metricTelemetryConsumers = new List<IMetricTelemetryConsumer>();
+        private List<ITraceTelemetryConsumer> traceTelemetryConsumers = new List<ITraceTelemetryConsumer>();
+
+        public IEnumerable<ITelemetryConsumer> TelemetryConsumers => this.consumers;
+
+        public void AddConsumers(IEnumerable<ITelemetryConsumer> newConsumers)
+        {
+            this.consumers = new List<ITelemetryConsumer>(this.consumers.Union(newConsumers));
+            this.metricTelemetryConsumers = this.consumers.OfType<IMetricTelemetryConsumer>().ToList();
```

Comment:
can we avoid allocating a new List here? Can we do` this.consumers.AddRange(newConsumers);` ?

Reply:
what's the purpose? I don't want this to ever get out of sync, so I thought this would be the easiest to future proof that. Why are you concerned with this allocation? This probably might happen once at startup, if it ever happens.

Reply:
this function is AddConsumers, not SetConsumers. So It can be called multiple times. I image the call pattern can be:
var telemetryManager  = new TelemetryManager();
telemetryManager.AddConsumers(newConsumers1);
telemetryManager.AddConsumes(newConsumesr2);
//inject into DI
services.AddSingleton<ITelemetryClient>(telemetryManager). 

User can call it like this, right?

Reply:
correct, but that's the point, we are nowhere near a hot or warm path, this is just a builder-like method. Even if you call it 100 times at startup and then never again, I wouldn't worry about those allocations. I prefer that if we add a RemoveConsumer, or the current dispose logic, that there is no chance we can screw up.
Also, now that you made me look closer, I did this mainly because I want an immutable copy of the list, so if there are any writes in progress while we update this, that we don't have race conditions that throw, and using a ConcurrentBag is not a good option for something that is not likely to ever change after startup.

Reply:
"this is not a hot or warm path " is not the reason to do additional allocation than needed. And  the reason of getting a immutable copy is not enough to handle race condition. You have potential to lose data when two are trying to AddConsumers at the same time. 

To handle race condition, you either make it a ConcurrentBag or remove the AddConsumer method entirely so that application can only set it through constructor. A half solution of using Union is not solving either problem of the two.

Reply:
We can address this in another PR or even in my logging PR since we want to merge this PR ASAP. but just state my opinion here.